### PR TITLE
Add new yaml-team-authorized-keys flag to provide authorized team key in one yaml file

### LINKED ISSUE #3551: Be able to specify team worker without the need of restart Concourse (TSA)
## What challenge are you facing?

As partially described in those both "related issues" :

  * https://github.com/concourse/concourse/issues/1105#issuecomment-473360524
  * https://github.com/concourse/concourse/issues/3549

I'm working with a mutualized Concourse server and dedicated workers (only) to teams. For each "users/customer" we create a dedicated team.
We don't provide default workers. A customer to run his jobs have to connect his own worker(s) to his team.

Let's say we have customer A and customer B.
Customer A will have a dedicated team A, on which he will assign Worker A to run his own jobs.

We want to prevent jobs of customer A running into Workers of customer B, and we also want to prevent Customer B to connect his workers on team A.

So to satisfy this issue, the current `tsa-team-authorized-keys` feature does perfectly the job.

BUT it has some drawbacks.

If we want to automate a bit the process, each time we create a new team in Concourse, we have to add a new param on the server and trigger a restart to take care of it.

eg 
```
--tsa-team-authorized-keys=teamA:/teamAfile
--tsa-team-authorized-keys=teamB:/teamBfile
....
--tsa-team-authorized-keys=teamN:/teamNfile
```

So if we create several teams by day, it becomes quickly hard to maintain, and creates potential downtime for customers.

**update** This would also improve k8s setup to put authorized keys in one configmap.

There ares 2 issues related to that :

  * 1) The need to add new command-line arguments for new Concourse team created.
  * 2) The need to restart Concourse when we add a key in those files (CF issue https://github.com/concourse/concourse/issues/1105#issuecomment-473360524)

## What would make this better?

I internally implemented a solution for 1) by adding a yaml authorized key flag :

```
--tsa-yaml-team-authorized-keys                              Path to file containing keys to authorize, in SSH authorized_keys yaml format. [$CONCOURSE_TSA_YAML_TEAM_AUTHORIZED_KEYS]
``` 

This flag targets a yaml file which contains authorized keys for each team in yaml format, following a bit this format https://bosh.io/jobs/tsa?source=github.com/concourse/concourse&version=4.2.1#p%3dteam_authorized_keys

```
  - team: teamA
    ssh_keys:
      - ssh-rsa keyA
  - team: teamB
    ssh_keys:
      - ssh-rsa keyB
```

So it solves the 2) issue, when I create a new Concourse team, I simply add a section for that team.
Then I tried the solution proposed in the issue for 1).

With the combination of those 2 fixes ( `1)` and `2)` ), I was able to automate the Concourse team creation with dedicated team worker.

My current workflow is now, create a concourse team, add a new section in the yaml to assign new worker key to that team. Then send a signal ( see `1)` ) to Concourse to trigger a refreshment of this updated yaml file.

This solution for my usecase get me rid of restarting concourse to assign new workers key to a team.

## Are you interested in implementing this yourself?

Yes, please give me your feedback/advises and I would be happy to submit a PR!

### PR DESCRIPTION
**Pull request by gaelL**
Add new yaml-team-authorized-keys flag to provide authorized team key in one yaml file

Fix #3551

Allow users to provide a yaml file containing all team authorized key following this format :

```
  - team: teamA
    ssh_keys:
      - ssh-rsa keyA
  - team: teamB
    ssh_keys:
      - ssh-rsa keyB
```

### CODE REVIEW

**Comment by gaelL** [comment]
Note : tests will fail as it require https://github.com/concourse/flag/pull/4 to be merged

**Comment by gaelL** [comment]
The requirement of flag pull request have been removed.

This pull request implement a `--team-authorized-keys-file` which currently work with a yaml file but we could add others format in the future.

**Comment by gaelL** [comment]
cc @clarafu this commit have been rebased.

To test it locally, I added a new `CONCOURSE_TSA_TEAM_AUTHORIZED_KEYS_FILE` envvar using docker compose pointing on a file I provided using `volumes` section on web container.

Make sure the yaml file content looks like 

```
  - team: teamA
    ssh_keys:
      - ssh-rsa keyA
  - team: teamB
    ssh_keys:
      - ssh-rsa keyB
```

As this code use `ssh.ParseAuthorizedKey` ssh pub keys have to be valid or they will be ignored.

Anyway you should see at the startup of the web container following lines of log : 

```
web_1     | {"timestamp":"2019-09-16T08:42:57.739212464Z","level":"info","source":"tsa","message":"tsa.load-team-authorized-keys-loaded","data":{"key":"ssh-rsa AAAAB3...","team":"teamA"}
```

**Comment by ddadlani** [comment]
Hi @gaelL, thanks for submitting this PR!

Just wondering, have you also considered watching a directory rather than a single file? Each file in the directory could correspond to a team and could contain the team's authorized keys. I'm curious as to whether you have gone down that path already, and what impact it would have on your workflow.

**Comment by gaelL** [comment]
Hi, @ddadlani I didn't considered to use a directory, I'm not sure what would be the best design for a directory behavior.
To give more visibility on my workflow.

I'm currently using one yaml file. As I have a production deployment in kubernetes,this file in a generated configmap from a third party software.
We dynamically generate Concourse team with associated workers and SSH keys.

To make it more efficient we implemented internally https://github.com/concourse/concourse/issues/1105 (I plan to provide a pull request for it after this one will be resolved).

So my workflow is basically : 
  * The team yaml file is generated automatically by a third party software
  * The content of this file is uploaded as kubernetes configMap
  * based on https://github.com/concourse/concourse/issues/1105 this thirdparty software will trigger a reload of this file to not have to restart Concourse to connect my new workers.

**Comment by Seraf** [comment]
Hello @ddadlani, do you have some feedbacks about this ? Thanks !

**Comment by ddadlani** [comment]
Hi @Seraf, I don't have anything more to add. Perhaps @vito or @concourse/runtime could take a look?

File: tsa/tsacmd/command.go [hunk 1]
```diff
@@ -29,9 +32,10 @@ type TSACommand struct {
...
-	AuthorizedKeys     flag.AuthorizedKeys            `long:"authorized-keys" description:"Path to file containing keys to authorize, in SSH authorized_keys format (one public key per line)."`
-	TeamAuthorizedKeys map[string]flag.AuthorizedKeys `long:"team-authorized-keys" value-name:"NAME:PATH" description:"Path to file containing keys to authorize, in SSH authorized_keys format (one public key per line)."`
+	HostKey                *flag.PrivateKey               `long:"host-key"        required:"true" description:"Path to private key to use for the SSH server."`
+	AuthorizedKeys         flag.AuthorizedKeys            `long:"authorized-keys" description:"Path to file containing keys to authorize, in SSH authorized_keys format (one public key per line)."`
+	TeamAuthorizedKeys     map[string]flag.AuthorizedKeys `long:"team-authorized-keys" value-name:"NAME:PATH" description:"Path to file containing keys to authorize, in SSH authorized_keys format (one public key per line)."`
+	TeamAuthorizedKeysFile flag.File                      `long:"team-authorized-keys-file" description:"Path to file containing keys to authorize, in SSH authorized_keys yaml format."`
```
[the remark this answers is not in the document]
Comment (gaelL) [reply tsa/tsacmd/command.go]: Nice catch I will commit your suggestion

**Comment by gaelL** [comment]
@vito 's change have been commited and rebased into one commit

**Pull request merged**